Modified launch files tutorial, added a note to the README.

Jakeisthesnake opened the pull request:
### Description

Launch File tutorial
Added explanations for the member function of MoveItConfigsBuilder.
Added note about launch nodes for custom .cpp files.
Having made these suggestions, I now wonder if I am being too verbose and just cluttering up the documentation when someone wanting to learn more can do what I did and read the source code. So maybe the launch file tutorial should remain unchanged.

README
Added a note that building MoveIt can be skipped if only adding explanations, note any executable code. This is mostly addressing my own pain of building MoveIt (about a 3-4 hour process) and I could see it stopping someone from adding a few helpful comments.


Please explain the changes you made, including a reference to the related issue if applicable

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] While waiting for someone to review your request, please consider reviewing [another open pull request](https://github.com/moveit/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

Jakeisthesnake committed:
Added to Launch Files tutorial explination of MoveitConfigsBuilder member funtions.
Added note to Launch FIles tutorial about nodes for custom cpp files.
Added note to README.md that building Moveit is not needed if only adding explinations, not exectuable code. This is mostly addressing the pain I feel of building MoveIt (3-4 hours on my laptop). Maybe I am an anomily.

Jakeisthesnake committed:
Corrected some spelling.

Jakeisthesnake committed:
Spelling corrections.

Jakeisthesnake commented:
As an aside, I would have liked to add more about '.planning_scene_monitor()' regarding when you need to set its parameters to be true, but I haven't found good explanations regarding common uses cases. If you can point me in the direction of where to learn more, I would be happy to add that info to the tutorial.

sea-bass commented on doc/how_to_guides/moveit_launch_files/moveit_launch_files_tutorial.rst:
```diff
@@ -213,3 +248,23 @@ Finally, we can tell our launch file to actually launch everything we described
             ]
         )
+
+Launching a custom .cpp file
+----------------------------
+
+While not part of the Getting Started tutorial, another common node to launch is one that executes a custom .cpp file:
```
A custom `.cpp` file doesn't sound technically precise -- I think what you may want to say is "a custom executable.

Also, this in isolation may not be entirely helpful if you also don't show a `CMakeLists.txt` snippet showing how to actually compile this executable.

But generally, there are other tutorials like https://moveit.picknik.ai/main/doc/tutorials/your_first_project/your_first_project.html which already cover this in full detail.

sea-bass commented on doc/how_to_guides/moveit_launch_files/moveit_launch_files_tutorial.rst:
```diff
@@ -213,3 +248,23 @@ Finally, we can tell our launch file to actually launch everything we described
+        package="custom_program_package",
+        executable="custom_program",
+        output="screen",
+        parameters=[your,
+                    parameters,
+                    here],
```
I wouldn't consider this too helpful because I always forget what the actual format of the parameters is -- it's a list of dictionaries or something? So I may consider using a real, valid input here instead of placeholder text.

sea-bass reviewed (commented):
Thank you for getting this started!

My first response here is that the information being added is largely already available in other tutorials. I would recommend reading through some of the other pages I've linked in the comments, and seeing if there's a better way to take advantage of cross-linking doc pages, but keeping each individual page focused on a specific subset of topics.

sea-bass commented:
> Ok, I'll make those edits. Do you know if any of the tutorials discuss the parameters for `.planning_scene_monitor()`? It seems lacking from the moveit_configuration_tutorial.

Thanks!

Maybe those could be mapped to the concepts page at https://moveit.picknik.ai/main/doc/concepts/planning_scene_monitor.html and/or https://moveit.picknik.ai/main/doc/examples/planning_scene_monitor/planning_scene_monitor_tutorial.html ?

Jakeisthesnake commented:
Yeah, I can start with those.

Jakeisthesnake commented:
If I wanted to cross link to the CMakeLists.txt file in the root moveit2_tutorials directory, how would I do that?
:doc:`moveit2_tutorials/CMakeLists.txt`
doesn't seem to work. I assume that is because :doc: only works for file in the doc folder?

sea-bass commented:
> If I wanted to cross link to the CMakeLists.txt file in the root moveit2_tutorials directory, how would I do that? :doc:`moveit2_tutorials/CMakeLists.txt` (the backticks turn "moveit2_tutorials/CMakeLists.txt" into inline code) doesn't seem to work. I assume that is because :doc: only works for files in the doc folder? I also assume that it would be poor form to directly link to "https://github.com/moveit/moveit2_tutorials/blob/main/CMakeLists.txt"

I believe you can use `:codedir:` -- you can see the definitions [here](https://github.com/moveit/moveit2_tutorials/blob/12edc7ea652a6590bda0fc0dc1cddfe8f0402748/conf.py#L189), and also search for similar usages in the `.rst` files in these docs.

There are other handy ones for linking to e.g. source code in the MoveIt 2 repo, etc.

Jakeisthesnake commented:
Thanks!

Jakeisthesnake commented:
I may be reading the definition wrong, but it looks like codedir can only reference stuff in the doc folder?
```
"codedir": (
        "https://github.com/"
        + html_context["github_user"]
        + "/moveit2_tutorials/blob/"
        + html_context["github_version"]
        + "doc/%s",
        "",
```

sea-bass commented:
It does look that way -- maybe instead of linking to the existing repo's `CMakeLists.txt`, you can just copy-paste a small relevant snippet of what you intend to show?

I assume this is just the compilation of a specific executable given source files?

Jakeisthesnake commented:
Yeah I could do that. Just walking through the CMakeLists.txt file for moveit_cpp tutorial.

Jakeisthesnake committed:
Moved most of the config specific edits to the configuration tutorial.
Added a link in the launch tutorial to the moveit_cpp tutorial.
Added the CMakeLists.txt code and expliantions to the moveit_cpp tutorial.

Jakeisthesnake committed:
Merge branch 'main' into main

Jakeisthesnake commented:
Updated the tutorial:
Moved most of the config specific edits to the configuration tutorial.
Aside from adding a link in the launch tutorial to the moveit_cpp tutorial, the launch tutorial went back to how it was.
Added the CMakeLists.txt code and explications to the moveit_cpp tutorial.

sea-bass commented on doc/examples/moveit_cpp/moveitcpp_tutorial.rst:
```diff
@@ -29,4 +29,50 @@ The entire code can be seen :codedir:`here in the MoveIt GitHub project<examples
+        )
+This node contains parameters which are passed to the executable associated with moveit_cpp_tutorial.
+
+The CMakeLists.txt File
+-----------------------
+During the build process, the CMakeLists.txt file assigns where the launcher will look for the executable.
+Below is the CMakeLists.txt file for this tutorial:
```
Add code font: ` ``CMakeLists.txt`` ` to all these mentions

Also "launcher" --> "launch file"

sea-bass commented on doc/examples/moveit_cpp/moveitcpp_tutorial.rst:
```diff
@@ -29,4 +29,50 @@ The entire code can be seen :codedir:`here in the MoveIt GitHub project<examples
+            executable="moveit_cpp_tutorial",
+            output="screen",
+            parameters=[moveit_config.to_dict()],
+        )
+This node contains parameters which are passed to the executable associated with moveit_cpp_tutorial.
```
Use ` ``code_font`` ` here for the tutorial name

sea-bass replied:
nit: Add a blank line in between these 2 -- although it seems to render fine as is

sea-bass commented on doc/examples/moveit_cpp/moveitcpp_tutorial.rst:
```diff
@@ -29,4 +29,50 @@ The entire code can be seen :codedir:`here in the MoveIt GitHub project<examples
+    DESTINATION share/${PROJECT_NAME}
+  )
+
+``add_executable`` builds the executable named moveit_cpp_tutorial from the source file ``src/moveit_cpp_tutorial.cpp``.
```
code font for the executable name here too

sea-bass commented on doc/examples/moveit_cpp/moveitcpp_tutorial.rst:
```diff
@@ -29,4 +29,50 @@ The entire code can be seen :codedir:`here in the MoveIt GitHub project<examples
+``PUBLIC`` makes the include directory available to targets that depend on ``moveit_cpp_tutorial``.
+
+``ament_target_dependencies`` specifies dependencies for the moveit_cpp_tutorial executable.
+``${THIS_PACKAGE_INCLUDE_DEPENDS}`` is a variable defined in the  ``moveit2_tutorials`` root directory's CMakeLists.txt file, which list common dependencies used in moveit2_tutorials.
```
code font for all the things here as well

sea-bass commented on doc/examples/moveit_cpp/moveitcpp_tutorial.rst:
```diff
@@ -29,4 +29,50 @@ The entire code can be seen :codedir:`here in the MoveIt GitHub project<examples
+``target_include_directories`` specifies the directories to search for header files during compilation.
+``PUBLIC`` makes the include directory available to targets that depend on ``moveit_cpp_tutorial``.
+
+``ament_target_dependencies`` specifies dependencies for the moveit_cpp_tutorial executable.
```
and here

sea-bass commented on doc/examples/moveit_cpp/moveitcpp_tutorial.rst:
```diff
@@ -29,4 +29,50 @@ The entire code can be seen :codedir:`here in the MoveIt GitHub project<examples
+
+``install(TARGET ...)`` and ``install(DIRECTORY ...)`` specify where to put the built executable and directories needed to run your program.
+
+For more details about custom executables, please browse the other examples on this site as well as the ROS documentation.
```
```suggestion
For more details about building executables in ROS packages, please browse the other examples on this site as well as the ROS documentation.
```

Also, consider maybe linking to the ROS documentation here?

sea-bass commented on doc/how_to_guides/moveit_configuration/moveit_configuration_tutorial.rst:
```diff
@@ -167,21 +167,68 @@ or you can include selected sub-components as follows:
+
+``MoveItConfigsBuilder`` (`defined here <https://github.com/moveit/moveit2/blob/main/moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py>`_) can take a few different types of arguments.
+
+* ``MoveItConfigsBuilder(package_name="package_name")``  will search for a package named "package_name".
```
code font on `"package_name"` as well

sea-bass commented on doc/how_to_guides/moveit_configuration/moveit_configuration_tutorial.rst:
```diff
@@ -167,21 +167,68 @@ or you can include selected sub-components as follows:
+As seen above, ``gen3`` is the robot name, and MoveIt looks for the package ``kinova_gen3_7dof_robotiq_2f_85_moveit_config`` instead of ``gen3_moveit_config``.
+
+``.robot_description`` can optionally take a file path to ``robot_name.urdf`` and/or assign a dictionary of argument mappings that are passed to the robot's urdf.xacro file.
+The file path to ``robot_name.urdf`` must be relative to your robot package, so if your robot package is ``/robot_package`` and the urdf (or urdf xacro) file is ``robot_package/config/robot_name.urdf``
+you would pass ``.robot_description(filepath="config/robot_name.urdf")``.
```
I think there may be a whitespace missing here

sea-bass commented on doc/how_to_guides/moveit_configuration/moveit_configuration_tutorial.rst:
```diff
@@ -167,21 +167,68 @@ or you can include selected sub-components as follows:
+Both arguments can be given, in which case the robot name is stored and the package with the explicitly given name will be loaded.
+As seen above, ``gen3`` is the robot name, and MoveIt looks for the package ``kinova_gen3_7dof_robotiq_2f_85_moveit_config`` instead of ``gen3_moveit_config``.
+
+``.robot_description`` can optionally take a file path to ``robot_name.urdf`` and/or assign a dictionary of argument mappings that are passed to the robot's urdf.xacro file.
+The file path to ``robot_name.urdf`` must be relative to your robot package, so if your robot package is ``/robot_package`` and the urdf (or urdf xacro) file is ``robot_package/config/robot_name.urdf``
+you would pass ``.robot_description(filepath="config/robot_name.urdf")``.
+If you don't provide a file path, but you do give ``MoveItConfigsBuilder`` a robot name (see above paragraph), MoveIt will look for ``robot_package/config/robot_name.urdf``.
+
+``.trajectory_execution`` loads trajectory execution and MoveIt controller manager's parameters from an optionally provided file path.
+If a file path isn't given, MoveIt looks for files in the package's ``config`` folder for files ending with ``controllers.yaml``.
+
+``.planning_scene_monitor`` allows you to set various parameters about what scene information is published and how often is it published.
+
+``.planning_pipelines`` allows to you to list the names of the planners you want available to be used by your robot.
+If you opt to not list pipelines, as in ``.planning_pipelines()``, MoveIt will look in the config folder of your package for files that end with "_planning.yaml".
+Additionally, if no pipelines are listed, MoveIt will load a set of planners from its own library - this can be disabled adding ``load_all=False`` as an argument to ``.planning_pipelines``.
+Listing the planner names specifiies which planners MoveIt should load; again these should be in your config folder.
+MoveIt will also pick one of your planners to be the default planner.
+If OMPL is one of your planners, it will be the default planner unless you set ``default_planning_pipeline`` to your desired default planner as in
+
+.. code-block:: python
+
+    .planning_pipelines(
+            pipelines=["ompl", "your_favorite_planner"],
+            default_planning_pipeline="your_favorite_planner",
+        )
+
+If OMPL is not in your planner list and you don't set a default planner, MoveIt will pick the first planner in the list.
```
Consider putting this in a bulleted list. Right now it renders like this which is a little dense to parse

![image](https://github.com/user-attachments/assets/6c0d6976-42a3-4309-8803-855ac3b36ea5)

sea-bass replied:
Consider linking to the OMPL Planner doc page here

sea-bass commented on doc/how_to_guides/moveit_launch_files/moveit_launch_files_tutorial.rst:
```diff
@@ -47,7 +47,7 @@ A handy way to refer to a MoveIt configuration package is to use the ``MoveItCon
 --------------------
 
-Once all the MoveIt configuration parameters have been loaded, you can launch the :ref:`Move Group Interface` using the entire set of loaded MoveIt parameters.
+Once all the MoveIt configuration parameters have been loaded, you can define the :ref:`Move Group Interface` Node using the entire set of loaded MoveIt parameters.
```
```suggestion
Once all the MoveIt configuration parameters have been loaded, you can define the :ref:`Move Group Interface` node using the entire set of loaded MoveIt parameters.

sea-bass commented on doc/how_to_guides/moveit_launch_files/moveit_launch_files_tutorial.rst:
```diff
@@ -213,3 +214,8 @@ Finally, we can tell our launch file to actually launch everything we described
             ]
         )
+
+Launching a custom executable
+----------------------------
+
+Later on these tutorials (:doc:`/doc/examples/moveit_cpp/moveit_cpp_tutorial`), you will learn how to create and launch a custom executable.
```
I would remove this.

sea-bass reviewed (commented):
Thanks for the fixes! I think this is looking good, as most of my remaining comments are small.
